### PR TITLE
OC-10413:Ability to schedule events for a list of participants

### PR DESCRIPTION
**Pull request by Taoli2018**
--- more changes focus on Async and write to log files

### CODE REVIEW

File: web/src/main/java/org/akaza/openclinica/controller/StudyEventController.java [hunk 1]
```diff
@@ -239,55 +256,89 @@ public RestfulServiceHelper getRestfulServiceHelper() {
...
+			 if(!(fileNm.endsWith(".csv")) ){
+				 throw new OpenClinicaSystemException("errorCode.notSupportedFileFormat","The file format is not supported at this time, please send CSV file, like *.csv ");
+			 }
+	
+			 int endIndex = fileNm.indexOf(".csv"); 
+			 String originalFileNm =  fileNm.substring(0, endIndex);
+			 logFileName = this.getRestfulServiceHelper().getMessageLogger().getLogfileNamewithTimeStamp(originalFileNm);
+			
+		}else {
+			throw new OpenClinicaSystemException("errorCode.emptyFile","The file is empty ");
```
Comment (yogi-oc) [inline web/src/main/java/org/akaza/openclinica/controller/StudyEventController.java]: Please log the exception before throwing. We don't know what happens after it is thrown.

File: web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java [hunk 2]
```diff
@@ -100,6 +103,70 @@ public void writeToLog(String subDir,String orginalFileName, String headerLine,S
...
+ public void writeToLog(String subDir,String orginalFileName, String headerLine,String msg,UserAccountBean userBean) {
+		
+    	BufferedWriter bw = null;
+		FileWriter fw = null;
+		boolean isNewFile = false;
+
+	    String logFileName;
+	   
+	    if(orginalFileName == null) {
+	    	;
```
Comment (yogi-oc) [inline web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java]: Log some message here.

File: web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java [hunk 3]
```diff
@@ -100,6 +103,70 @@ public void writeToLog(String subDir,String orginalFileName, String headerLine,S
...
+	 			
+	 			if(msg != null) {
+	 				bw.write(msg);	
+	 				bw.write("\n");
+	 			}	
+	 			
+	 			bw.close();						
+	 	       
+	 	    } catch (Exception e) {
+	 	        e.printStackTrace();
```
Comment (yogi-oc) [inline web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java]: Do not print the trace. Log it. 
Do not eat the exception. Throw it back.

File: web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java [hunk 4]
```diff
@@ -100,6 +103,70 @@ public void writeToLog(String subDir,String orginalFileName, String headerLine,S
...
+	 	    }finally {
+	 			try {
+	 				if (bw != null)
+	 					bw.close();
+	 				if (fw != null)
+	 					fw.close();
+	 			} catch (IOException ex) {
+	 				ex.printStackTrace();
```
Comment (yogi-oc) [inline web/src/main/java/org/akaza/openclinica/controller/helper/MessageLogger.java]: same as above. Log it and throw it back.

File: web/src/main/java/org/akaza/openclinica/service/StudyEventServiceImpl.java [hunk 5]
```diff
@@ -260,6 +260,216 @@ public RestReponseDTO scheduleStudyEvent(HttpServletRequest request, String stud
...
+	    	
+	    	// check study first
+	    	currentStudy = studyDao.findStudyByOid(studyOID);
+	    	
+	    	if(currentStudy == null) {
+	    		errMsg = "The study {" + studyOID + "} is not existing in the system.";
+	    		throw new OpenClinicaException(errMsg,ErrorConstants.ERR_STUDY_NOT_EXIST);
+	    	}else if (currentStudy.getStatus().equals(Status.LOCKED)) {
+	    		errMsg = "The study {" + studyOID +"} has been LOCKED.";
+	    		throw new OpenClinicaException(errMsg,ErrorConstants.ERR_STUDY_LOCKED);
```
Comment (yogi-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyEventServiceImpl.java]: Log the error

**Comment by Taoli2018** [comment]
did all changes based on code review,please review again. Thanks

**Comment by yogi-oc** [comment]
:+1:

**Pull request merged**